torch/distributed/_shard/checkpoint/resharding.py

```diff
@@ -0,0 +1,290 @@
+        (dimension, `saved_shard` offset, `current_shard` offset, length)
+
+    Offsets are relative to each shard.
+    """
+    narrows = []
+    for dim, (
+        saved_shard_offset,
+        current_shard_offset,
+        saved_shard_size,
+        current_shard_size,
+    ) in enumerate(
+        zip(
+            saved_shard.shard_offsets,
+            current_shard.shard_offsets,
+            saved_shard.shard_sizes,
+            current_shard.shard_sizes,
+        )
+    ):
+        min_range_end = min(
+            saved_shard_offset + saved_shard_size,
+            current_shard_offset + current_shard_size,
+        )
+
+        length = min_range_end - max(current_shard_offset, saved_shard_offset)
+
+        if saved_shard_offset > current_shard_offset:
+            offset_for_saved_tensor = 0
+            offset_for_current_tensor = saved_shard_offset - current_shard_offset
+        else:
+            offset_for_saved_tensor = current_shard_offset - saved_shard_offset
+            offset_for_current_tensor = 0
+
+        narrows.append(
+            (dim, offset_for_saved_tensor, offset_for_current_tensor, length)
+        )
+
+    return narrows
+
+
+def _get_sharded_tensor_element_size(tensor: ShardedTensor) -> int:
+    if len(tensor.local_shards()) > 0:
+        test_tensor = tensor.local_shards()[0].tensor
+    else:
+        dtype = tensor.metadata().tensor_properties.dtype
+        test_tensor = torch.empty((1,), dtype=dtype)
+
+    return test_tensor.element_size()
+
+
+def _compute_sharded_tensor_md(
+    tensor: ShardedTensor,
+    shard_to_storage_key: Dict[str, str]
+) -> ShardedTensorStorageMetadata:
+    smd = []
+    for shard_md in tensor.metadata().shards_metadata:
+        shard_storage_key = shard_to_storage_key[_get_shard_key(shard_md)]
+
+        shard_size = 1
+        for d in shard_md.shard_sizes:
+            shard_size *= d
+
+        # not particularly great
+        storage_size = shard_size * _get_sharded_tensor_element_size(tensor)
+
+        one_smd = ShardStorageMetadata(
+            shard_metadata=shard_md,
+            storage_key=shard_storage_key,
+            length=storage_size,
+        )
+        smd.append(one_smd)
+
+    return ShardedTensorStorageMetadata(
+        tensor_metadata=tensor.metadata(),
+        storage_metadata=smd,
+    )
+
+
+def _get_shard_key(shard: ShardMetadata) -> str:
+    """
+    Compute an unique key for a shard.
+
+    This key is unique vis-a-vis other shard of the owning ShardedTensor
+    """
+    return "_".join([str(i) for i in shard.shard_offsets])
+
+def _get_shard_storage_key(
+    tensor_storage_key: str,
+    shard: ShardMetadata,
+    storage_key_to_fqn: Dict[str, str]
+) -> str:
+    shard_key = f"{tensor_storage_key}{STORAGE_KEY_SEPARATOR}{_get_shard_key(shard)}"
+
+    return _create_storage_key(storage_key_to_fqn, shard_key)
+
+
+def _prepare_sharded_tensor_write(
+    sharded_tensor: ShardedTensor,
+    storage_key: str,
```

Comment:
nit: This should be `fqn`? It was a bit confusing while reading the code, where I thought this is the storage_key in the StorageMetadata but it is actually the FQN of the parameter?